### PR TITLE
Havardpe/use string ids as tensor labels

### PR DESCRIPTION
**Pull request by havardpe**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@arnej27959 please review

### CODE REVIEW

**Review by arnej27959 — APPROVED** [review]
looks good, perf tests show significant improvements.

File: vespalib/src/vespa/vespalib/util/shared_string_repo.cpp [hunk 1]
```diff
@@ -6,6 +6,18 @@ namespace vespalib {
 
 SharedStringRepo::Partition::~Partition() = default;
 
+void
+SharedStringRepo::Partition::find_leaked_entries(size_t my_idx) const
+{
+    for (size_t i = 0; i < _entries.size(); ++i) {
+        if (!_entries[i].is_free()) {
+            size_t id = (((i << PART_BITS) | my_idx) + 1);
+            fprintf(stderr, "WARNING: shared_string_repo: leaked string id: %zu ('%s')\n",
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/util/shared_string_repo.cpp:15]: Prefer using LOG() instead.

File: vespalib/src/vespa/vespalib/util/shared_string_repo.h [hunk 2]
```diff
@@ -34,21 +35,43 @@ class SharedStringRepo {
...
-                str = key.str;
+                uint32_t next = _hash;
+                _hash = key.hash;
+                _ref_cnt = 1;
+                _str = key.str;
                 return next;
             }
             void fini(uint32_t next) {
-                ref_cnt = next;
+                _hash = next;
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/util/shared_string_repo.h:59]: The name of this function is not obvious to me.

File: vespalib/src/vespa/vespalib/util/shared_string_repo.cpp (hunk 1 above)
Comment (havardpe) [reply vespalib/src/vespa/vespalib/util/shared_string_repo.cpp:15]: Can we guarantee that the logging pipeline is still intact at this point?

Comment (geirst) [reply vespalib/src/vespa/vespalib/util/shared_string_repo.cpp:15]: @arnej27959 / @baldersheim / @toregge : Any comments on that?

File: vespalib/src/vespa/vespalib/util/shared_string_repo.h (hunk 2 above)
Comment (havardpe) [reply vespalib/src/vespa/vespalib/util/shared_string_repo.h:59]: init/fini is like ctor/dtor

File: vespalib/src/vespa/vespalib/util/shared_string_repo.cpp (hunk 1 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/util/shared_string_repo.cpp:15]: I am not sure what c++ says abut destruction order of static versus global objects

**Pull request merged**